Add first version of the guide for writing REST microservices using virtual threads (#27840)
quarkusio/quarkus: docs/src/main/asciidoc/quarkus-virtual-threads.adoc

:blockingannotation: https://javadoc.io/doc/io.smallrye.common/smallrye-common-annotation/latest/io/smallrye/common/annotation/Blocking.html
:vthreadjep: https://openjdk.org/jeps/425
:thread: https://docs.oracle.com/en/java/javase/18/docs/api/java.base/java/lang/Thread.html
:mutiny-vertx-sql: https://smallrye.io/smallrye-mutiny-vertx-bindings/2.26.0/apidocs/io/vertx/mutiny/sqlclient/package-summary.html
:pgsql-driver: https://javadoc.io/doc/org.postgresql/postgresql/latest/index.html

This guide explains how to benefit from Java 19 virtual threads when writing REST services in Quarkus.

[TIP]
====
This is the reference guide for using virtual threads to write reactive REST services.
Please refer to the xref:rest-json.adoc[Writing JSON REST services guides] for a lightweight introduction to
reactive REST services and to the xref:resteasy-reactive.adoc[Writing REST Services with RESTEasy Reactive] guide for a
detailed presentation.
====

== What are virtual threads ?

=== Terminology
OS thread::
A "thread-like" data-structure managed by the Operating System.

Platform thread::

> Until Java 19, every instance of the link:{thread}[Thread] class was a platform thread, that is, a wrapper around an OS

Reviewer: ```suggestion
Up until Java 19, every instance of the link:{thread}[Thread] class was a platform thread, that is, a wrapper around an OS
```

> thread. Creating a platform creates an OS thread, blocking a platform thread blocks an OS thread.

Reviewer: Is there something missing after `platform`? `thread`?

Author: indeed it was thread

Virtual thread::

> Lightweight, JVM-managed threads. They extend the link:{thread}[Thread] class but are not tied to one specific OS thread.

Reviewer 2: Add: Scheduling virtual threads is the responsability of the JVM.

Carrier thread::
A platform thread used to execute a virtual thread is called a carrier.

> This isn't a class distinct from link:{Thread}[Thread] or VirtualThread but rather a functional denomination.

Reviewer: I don't think `VirtualThread` has been mentioned yet, right?

Author: just above, it is a quick definition before we go further in detailing what they are

> === The differences between virtual threads and platform threads

Reviewer: ```suggestion
=== Differences between virtual threads and platform threads
```

We will give a brief overview of the topic here, please refer to the link:{vthreadjep}[JEP 425] for more information.

> Virtual threads a feature available since Java 19 aiming at providing a cheap alternative to platform threads for

Reviewer: ```suggestion
Virtual threads are a feature available since Java 19 aiming at providing a cheap alternative to platform threads for
```

> IO-bound workloads.

Reviewer 2: try to use one sentence per line.

> Until now, platform threads were the concurrency unit of the JVM. They are a thin wrapper over OS structures.

Reviewer 2: ```suggestion
Until now, platform threads were the concurrency unit of the JVM. They are a wrapper over OS structures.
```

This means that creating a Java platform thread actually results in creating a "thread-like" structure in your operating
system.

Virtual threads on the other hand are managed by the JVM. In order to be executed, they need to be mounted on a platform

> thread (which acts as a carrier).

Reviewer 2: ```suggestion
thread (which acts as a carrier, to that thread is a carrier thread).
```

As such, they have been designed to offer the following characteristics:

Lightweight :: Virtual threads occupy less space than platform threads in memory.

> Hence, it becomes possible to use more virtual threads than platform threads simultaneously before blowing up the heap.

Reviewer: ```suggestion
Hence, it becomes possible to use more virtual threads than platform threads simultaneously without blowing up the heap.
```

> By default, platform threads are created with a stack of about 1 MB (need ref) where virtual threads stack is

Reviewer: ```suggestion
By default, platform threads are created with a stack of about 1 MB (need ref) where virtual threads stack is
```

The `need` ref part should be fixed :)

> By default, platform threads are created with a stack of about 1 MB (need ref) where virtual threads stack is

Reviewer 2: ```suggestion
By default, platform threads are created with a stack of about ~1 MB where virtual threads stack is
```

Author: fixed it too, added a ref to a presentation by Ron Pressler

"pay-as-you-go".

Cheap to create:: Creating a platform thread in Java takes time.
Currently, techniques such as pooling where threads are created once then reused are strongly encouraged to minimize the
time lost in starting them (as well as limiting the maximum number of threads to keep memory consumption low).

> Those techniques are now useless: virtual threads are supposed to be disposable entities that we create when we need

Reviewer: ```suggestion
Those techniques are now outdated: virtual threads are supposed to be disposable entities that we create when we need
```

> Those techniques are now outdated: virtual threads are supposed to be disposable entities that we create when we need

Reviewer 2: ```suggestion
Virtual threads are supposed to be disposable entities that we create when we need
```

them, it is discouraged to pool them or to reuse them for different tasks.

> Cheap to block:: When performing blocking IO, the underlying OS thread wrapped by the Java platform thread is put in a

Reviewer 2: ```suggestion
Cheap to block:: When performing blocking I/O, the underlying OS thread wrapped by the Java platform thread is put in a
```

wait queue and a context switch occurs to load a new thread context onto the CPU core. This operation takes time. Since
virtual threads are managed by the JVM, no underlying OS thread is blocked when they perform a blocking operation. Their
state is simply stored in the heap and another Virtual thread is executed on the same Java platform thread.

=== Virtual threads are useful for IO-bound workloads only
We now know that we can create way more virtual threads than platform threads. One could be tempted to use virtual
threads to perform long computations (CPU-bound workload). This is useless if not counterproductive.
CPU-bound doesn't consist in quickly swapping threads while they need to wait for the completion of an IO but in leaving
them attached to a CPU-core to actually compute something. In this scenario, it is useless to have thousands of threads
if we have tens of CPU-cores, virtual threads won't enhance the performance of CPU-bound workloads.

== Bringing virtual threads to reactive REST services

> Since virtual threads are disposable entities, the fundamental idea of quarkus-loom is to offload the execution of an

Reviewer 2: we need a new name. quarkus-loom is not great.
What about the Quarkus Virtual Thread feature or support?

Reviewer 2: go with it.

xref:resteasy-reactive.adoc#execution-model-blocking-non-blocking[Execution model, blocking, non-blocking] for more
information.

=== Getting started

Add the following import to your build file:

[source,xml,role="primary asciidoc-tabs-target-sync-cli asciidoc-tabs-target-sync-maven"]
.pom.xml
----
<dependency>
    <groupId>io.quarkus</groupId>
    <artifactId>quarkus-resteasy-reactive</artifactId>
</dependency>
----

[source,gradle,role="secondary asciidoc-tabs-target-sync-gradle"]
.build.gradle
----
implementation("io.quarkus:quarkus-resteasy-reactive")
----

You also need to make sure that you are using the version 19 of Java, this can be enforced in your pom.xml file
with the following:

[source,xml,role="primary asciidoc-tabs-target-sync-cli asciidoc-tabs-target-sync-maven"]
.pom.xml
----
<properties>
    <maven.compiler.source>19</maven.compiler.source>
    <maven.compiler.target>19</maven.compiler.target>
</properties>
----

Virtual threads are still an experimental feature, you need to start your application with the `--enable-preview` flag:

[source, bash]
----
java --enable-preview -jar target/quarkus-app/quarkus-run.jar

> ----

Reviewer: The step about building the application has been skipped. Is there anything special users need to do?

Author: not anymore, they only need to worry about the building step if they use the netty-loom extension, this is detailed in the part about the extension

Author: not sure what you mean @cescoffier ?

Reviewer 2: The dev mode required a -Dloom-open

Author: It doesn't anymore, it does if you use the Netty-loom extension though but there is a part about it in the guide

> Quarkus Dev

Reviewer 2: header? or something else is missing?

The example below shows the differences between three endpoints, all of them querying a fortune in the database then
returning it to the client.

- the first one uses the traditional blocking style, it is considered blocking due to its signature.
- the second one uses Mutiny reactive streams in a declarative style, it is considered non-blocking due to its signature.
- the third one uses Mutiny reactive streams in a synchronous way, since it doesn't return a "reactive type" it is
considered blocking and the link:{runonvthread}[@RunOnVirtualThread] annotation can be used.

When using Mutiny, alternative "xAndAwait" methods are provided to be used with virtual threads. They ensure that

> waiting for the completion of the IO will not "pin" the carrier thread and deteriorate performance.

Reviewer 2: ```suggestion
waiting for the completion of the I/O will not "pin" the carrier thread and deteriorate performance.
```

Also pinning has not been explained.

Reviewer 2: you can add a ref pointing to the following section.

            }
            return todoList;
        });
    }

    @GET
    @RunOnVirtualThread
    @Path("/quoted-virtual-thread")
    public List<Fortune> getAllQuotedBlocking() {
        //we get the list of fortunes
        var fortunes = repository.findAllAsyncAndAwait();

        //we get the list of quotes
        var quotes = getQuotes(fortunes.size()).await().indefinitely();

        //we append each quote to each fortune
        for(int i=0; i  < fortunes.size();i ++){
            fortunes.get(i).title+= "   -  "+quotes.get(i);
        }
        return todos;
    }

}
----

== Pinning cases
The notion of "cheap blocking" might not always be true: in certain occasions a virtual thread might "pin" its carrier
(the platform thread it is mounted upon).
In this situation, the platform thread is blocked exactly as it would have been in a typical blocking scenario.

According to link:{vthreadjep}[JEP 425] this can happen in two situations:

- when a virtual thread executes performs a blocking operation inside a `synchronized` block or method
- when it executes a blocking operation inside a native method or a foreign function.

> It can be fairly easy to avoid these situations in our own code, but it is hard to verify every extension we use.

Reviewer: ```suggestion
It can be fairly easy to avoid these situations in our own code, but it is hard to verify every dependency we use.
```

> Typically, while experimenting with Quarkus-virtual-threads, we realized that using the

Reviewer 2: ah, another name?

Author: removed it, too heavy

advise you stick to using reactive constructs.

This degradation of performance doesn't seem to come from virtual threads themselves but from the interactions between
Vert.x/Netty (Quarkus underlying reactive engine) and the virtual threads.
This was illustrated in the issue that we will now describe.

=== The Netty problem
For JSON serialization, Netty uses their custom implementation of thread locals, `FastThreadLocal` to store buffers.
When using virtual threads in quarkus, then number of virtual threads simultaneously living in the service is directly
related to the incoming traffic. It is possible to get hundreds of thousands, if not millions, of them.

If they need to serialize some data to JSON they will end up creating as many instances of `FastThreadLocal`, resulting
on a massive memory consumption as well as exacerbated pressure on the garbage collector.
This will eventually affect the performance of the application and inhibit its scalability.

This is a perfect example of the mismatch between the reactive stack and the virtual threads. The fundamental hypothesis

> are completely different and result in different optimization. Netty expects a system using few event-loops (twice the

Reviewer: ```suggestion
are completely different and result in different optimizations. Netty expects a system using few event-loops (twice the
```

Reviewer: Also, quarkus now defaults to exactly the number of cores when determining the event loop count, not double that used to be the case.

number of CPU cores by default in Quarkus), but it gets hundreds of thousands of threads.

> 

Reviewer 2: Add link to my email.

Author: mentioned it this way

=== Our solution to the Netty problem
In order to avoid this wasting of resource without modifying Netty upstream, we wrote an extension that modifies the
bytecode of the class responsible for creating the thread locals at build time.
Using this extension, performance of quarkus-virtual-threads for the Json Serialization test of the Techempower suite
increased by nearly 80%, making quarkus-virtual-threads almost as good as quarkus-reactive.

To use it, it needs to be added as a dependency:

[source,xml,role="primary asciidoc-tabs-target-sync-cli asciidoc-tabs-target-sync-maven"]
.pom.xml
----
<dependency>
    <groupId>io.quarkus</groupId>

>     <artifactId>quarkus-netty-loom-adaptor</artifactId>

Reviewer 2: does it work again?

Author: always did, I just got the flags wrong when I tried the demo last time

</dependency>
----

Furthermore, some operations undertaken by this extension need special access, it is necessary to

- compile the application with the flag `-Dnet.bytebuddy.experimental`

> - open the `java.base.lang` module at runtime with the flag `--add-opens java.base/java.lang=ALL-UNNAMED`

Reviewer: We should show an example of this so users can just copy and paste

Reviewer: Nevermind, I see it's added below
